File: app/src/main/java/vinova/henry/com/hotfilm/feature/home/MovieAdapter.java
package vinova.henry.com.hotfilm.feature.home;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.LoaderManager;
import android.content.ClipData;
import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.CardView;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import vinova.henry.com.hotfilm.R;
import vinova.henry.com.hotfilm.feature.detail.DetailActivity;
import vinova.henry.com.hotfilm.models.Movie;

public class MovieAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final int VIEW_TYPE_ITEM = 0;
    private final int VIEW_TYPE_LOADING = 1;
    private IOnLoadMoreListener onLoadMoreListener;
    private boolean isLoading;
    private Activity activity;
    private Context mContext;
    private List<Movie> movies;
    private int visibleThreshold = 5;
    private int lastVisibleItem, totalItemCount;

    MovieAdapter(RecyclerView recyclerView, Context mContext, Activity activity) {
        this.activity = activity;
        this.mContext = mContext;
        movies = new ArrayList<>();

        final GridLayoutManager layoutManager = (GridLayoutManager) recyclerView.getLayoutManager();
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                totalItemCount = layoutManager.getItemCount();
                lastVisibleItem = layoutManager.findLastVisibleItemPosition();
                if (!isLoading && totalItemCount <= (lastVisibleItem + visibleThreshold)) {
                    if (onLoadMoreListener != null) {
                        onLoadMoreListener.onLoadMore();
                    }
                    isLoading = true;
                }
            }
        });
    }

    @Override
    public int getItemViewType(int position) {
        return movies.get(position) == null ? VIEW_TYPE_LOADING : VIEW_TYPE_ITEM;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE_ITEM){
            View v = LayoutInflater.from(mContext).inflate(R.layout.item_movie_cardview, parent, false);
            return new ItemViewHolder(v);
        } else if (viewType == VIEW_TYPE_LOADING){
            View v = LayoutInflater.from(mContext).inflate(R.layout.item_loading, parent, false);
            return new LoadingViewHolder(v);
        } else return null;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, @SuppressLint("RecyclerView") final int position) {

        if (holder instanceof ItemViewHolder){
            ItemViewHolder viewHolder = (ItemViewHolder) holder;
            String poster = "https://image.tmdb.org/t/p/w500/" + movies.get(position).getPosterPath();
            Picasso.with(mContext)
                    .load(poster)
                    .into(viewHolder.imBackdrop);
            viewHolder.tvTitle.setText(movies.get(position).getTitle());

            viewHolder.cvMovie.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent intent = new Intent(mContext, DetailActivity.class);
                    intent.putExtra("Movie", (Serializable) movies.get(position));
                    mContext.startActivity(intent);
                }
            });
        } else if (holder instanceof LoadingViewHolder){
            LoadingViewHolder loadingViewHolder = (LoadingViewHolder) holder;
            loadingViewHolder.progressBar.setIndeterminate(true);
        }

    }

    @Override
    public int getItemCount() {
        return movies == null ? 0 : movies.size();
    }

    void setOnLoadMoreListener(IOnLoadMoreListener mOnLoadMoreListener) {
        this.onLoadMoreListener = mOnLoadMoreListener;
    }

    void setMovies(List<Movie> movies) {
        this.movies = movies;
    }

    void setLoaded() {
        isLoading = false;
    }

    class ItemViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.im_backdrop)
        ImageView imBackdrop;
        @BindView(R.id.tv_title)
        TextView tvTitle;
        @BindView(R.id.cv_movie)
        CardView cvMovie;

        ItemViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
        }
    }
    class LoadingViewHolder extends RecyclerView.ViewHolder{

        @BindView(R.id.progressBar1)
        ProgressBar progressBar;
        public LoadingViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/vinova/henry/com/hotfilm/feature/home/MovieAdapter.kt
package vinova.henry.com.hotfilm.feature.home

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.support.v7.widget.CardView
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView

import com.squareup.picasso.Picasso

import java.util.ArrayList

import butterknife.BindView
import butterknife.ButterKnife
import vinova.henry.com.hotfilm.R
import vinova.henry.com.hotfilm.feature.detail.DetailActivity
import vinova.henry.com.hotfilm.models.Movie

class MovieAdapter internal constructor(recyclerView: RecyclerView, private val mContext: Context) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val VIEW_TYPE_ITEM = 0
    private val VIEW_TYPE_LOADING = 1
    private lateinit var onLoadMoreListener: IOnLoadMoreListener
    private var isLoading: Boolean = false
    private var movies: List<Movie>? = null
    private val visibleThreshold = 5
    private var lastVisibleItem: Int = 0
    private var totalItemCount: Int = 0

    init {
        movies = ArrayList()

        val layoutManager = recyclerView.layoutManager as GridLayoutManager
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                totalItemCount = layoutManager.itemCount
                lastVisibleItem = layoutManager.findLastVisibleItemPosition()
                if (!isLoading && totalItemCount <= lastVisibleItem + visibleThreshold) {
                    if (onLoadMoreListener != null) {
                        onLoadMoreListener!!.onLoadMore()
                    }
                    isLoading = true
                }
            }
        })
    }

    override fun getItemViewType(position: Int): Int {
        return if (movies!![position] == null) VIEW_TYPE_LOADING else VIEW_TYPE_ITEM
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        if (viewType == VIEW_TYPE_ITEM) {
            val v = LayoutInflater.from(mContext).inflate(R.layout.item_movie_cardview, parent, false)
            return ItemViewHolder(v)
        } else  {
            val v = LayoutInflater.from(mContext).inflate(R.layout.item_loading, parent, false)
            return LoadingViewHolder(v)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, @SuppressLint("RecyclerView") position: Int) {

        if (holder is ItemViewHolder) {
            val poster = "https://image.tmdb.org/t/p/w500/" + movies!![position].posterPath
            Picasso.with(mContext)
                    .load(poster)
                    .into(holder.imBackdrop)
            holder.tvTitle!!.text = movies!![position].title

            holder.cvMovie!!.setOnClickListener {
                val intent = Intent(mContext, DetailActivity::class.java)
                intent.putExtra("Movie", movies!![position])
                mContext.startActivity(intent)
            }
        } else if (holder is LoadingViewHolder) {
            holder.progressBar!!.isIndeterminate = true
        }

    }

    override fun getItemCount(): Int {
        return if (movies == null) 0 else movies!!.size
    }

    internal fun setOnLoadMoreListener(mOnLoadMoreListener: IOnLoadMoreListener, function: () -> Boolean) {
        this.onLoadMoreListener = mOnLoadMoreListener
    }

    internal fun setMovies(movies: List<Movie>) {
        this.movies = movies
    }

    internal fun setLoaded() {
        isLoading = false
    }

    internal inner class ItemViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        @BindView(R.id.im_backdrop)
        var imBackdrop: ImageView? = null
        @BindView(R.id.tv_title)
        var tvTitle: TextView? = null
        @BindView(R.id.cv_movie)
        var cvMovie: CardView? = null

        init {
            ButterKnife.bind(this, view)
        }
    }

    internal inner class LoadingViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        @BindView(R.id.progressBar1)
        var progressBar: ProgressBar? = null

        init {
            ButterKnife.bind(this, itemView)
        }
    }
}
